Updated surface flux tests and docs. (#1328)
CliMA/ClimateMachine.jl: docs/src/Theory/Common/SurfaceFluxes.md

> <em>Internally abbreviated as GLGS</em>

>Hill, R.J., 1989. Implications of Monin–Obukhov Similarity Theory for Scalar Quantities. _J. Atmos. Sci.,_  **46**, 2236–2244, [https://doi.org/10.1175/1520-0469(1989)046<2236:IOMSTF>2.0.CO;2](https://doi.org/10.1175/1520-0469(1989)046%3C2236:IOMSTF%3E2.0.CO;2) 

### Parameters
* $L_{MO}$ = Monin-Obukhov length
* $X_{\star}$ = Scale parameter for variable $X$
* $X_{sfc}$ = Surface value of variable $X$ 
* $\kappa$ = von-Karman coefficient (typically $0.4$)
* $\mathrm{D_T} = \nu/Pr_{T}$ is the diffusivity
*  $Pr_{T}$ the neutral limit turbulent Prandtl number (typically $\frac{1}{3}$)
* $\nu$ the kinematic viscosity
* $w$, the wall normal velocity component
* $\Delta z$ = Model reference altitude
* $\delta_{\chi_{i}}$ a measure of mixing efficiency for $i$th tracer 
* $\zeta = z/L_{MO}$ similarity variable
* $\nu/u_{\star}$ Viscous lengthscale

### Fluxes

* Momentum fluxes due to boundary shear
```math 
-\rho \tau_{j3}|_{\mathrm{sfc}} = \rho \overline{w^{\prime}u_{j}^{\prime}}|_{\mathrm{sfc}} 
``` where $u$ represents the surface parallel velocity component, for $j = 1,2$

> *  Sum of specific enthalpy and moist static energy flux

Reviewer: This is just the moist static energy flux (MSE contains specific enthalpy plus potential energy)

```math 
\rho\mathrm{d}h_{t}|_{\mathrm{sfc}} = \rho \mathrm{(J+D)|_{sfc}} = -\rho D_{T}\nabla h_{t}|_{\mathrm{sfc}} = \rho[c_{pm}\overline{w^{\prime}\theta^{\prime}} + L_{v}\overline{w^{\prime}q_{t}^{\prime}}]|_{\mathrm{sfc}} 
```

* Subgrid flux of total water specific humidity
```math 

> \rho\mathrm{d}q_{t}|_{\mathrm{sfc}}  = -\rho D_{T}\nabla q_{t}|_{\mathrm{sfc}}  = \rho L_{v} \overline{w^{\prime}q_{t}^{\prime}}|_{\mathrm{sfc}} 

Reviewer: ```suggestion
\rho\mathrm{d}q_{t}|_{\mathrm{sfc}}  = -\rho D_{T}\nabla q_{t}|_{\mathrm{sfc}}  = \rho \overline{w^{\prime}q_{t}^{\prime}}|_{\mathrm{sfc}} 
```

```

> * Subgrid flux of tracer $i$, for $i = 1..n$ tracers, n an integer

Reviewer: ```suggestion
* Subgrid flux of tracer $i$, for $i = 1, \dots, n$ tracers, n an integer
```

```

We also have, by assumption that each variable can be represented in terms of the universal functions $\Phi_{k}$ for the $k$th variable,
* $\overline{w^{\prime}u^{\prime}}|_{\mathrm{sfc}} = -u_{\star}^2$
* $\overline{w^{\prime}\theta^{\prime}}|_{\mathrm{sfc}} = -u_{\star}\theta_{\star} = \frac{J}{\rho c_{pm}}$
* $\overline{w^{\prime}q_{t}^{\prime}}|_{\mathrm{sfc}} = -u_{\star}q_{t\star} = \frac{D}{\rho L_{v}}$
* $\overline{w^{\prime}\chi_{i}^{\prime}}|_{\mathrm{sfc}} = -u_{\star}\chi_{i\star}$

### Boundary conditions
Required variables for boundary condition prescription in `ClimateMachine` are

#### Momentum 
$u_{\star}$  in `bc_momentum.jl` . Usage is through the `DragLaw` function. Current use cases include prescribed values for $u_{\star}$, or prescribed exchange coefficients.

#### Energy
$\hat{n} \cdot \mathrm{d}h_{tot}$  in `bc_energy.jl` . Usage currently is through the `PrescribedEnergyFlux`

#### Moisture 
$\hat{n} \cdot \mathrm{d}q_{tot}$   in `bc_moisture.jl`  . Usage currently is through `PrescribedMoistureFlux`

#### Tracer Flux 

> $\hat{n} \cdot \mathrm{d}\chi_{i}$   in `bc_moisture.jl`  . No use cases currently. Tracers do not permeate through boundaries in present implementation. 

Reviewer: ```suggestion
$\hat{n} \cdot \mathrm{d}\chi_{i}$   in `bc_tracers.jl`  . No use cases currently. Tracers do not permeate through boundaries in present implementation. 
```
(and add a placeholder `bc_tracers.jl`

In each of these cases $\hat{n}$ corresponds to the upward pointing unit-normal vector. 

## Equations

> We start by defining the universal functions for the velocity, enthalpy and moisture terms. _NK_ express the Monin-Obukhov length in terms of the potential temperature 

Reviewer: ```suggestion
We start by defining the universal functions for the velocity, enthalpy and moisture terms. _NK_ express the Monin-Obukhov length in terms of the virtual dry static energy 
```

```math 
L_{MO} = -\frac{u_{\star}^3 \overline{\theta}}{\kappa g \overline{w^{\prime}\theta^{\prime}}|_{sfc}} 
```

From the above definitions for the fluxes in the moist, compressible flow equations, we substitute the potential temperature flux $\overline{w^{\prime}\theta^{\prime}} = J/(\rho c_{pm})$, $\overline{w^{\prime}u^{\prime}} = \tau/\rho)$ to recover eq (1) in _GLGS_. 

```math 
L_{MO} = -\frac{(\tau/\rho)^{(3/2)} \overline{\theta}}{\kappa g \frac{J}{(\rho c_{pm})}|_{sfc}} 
```

Given the formulation of the compressible flow equations in `ClimateMachine`  we arrive at the following formulation for the Monin-Obukhov similarity length

```math 
L_{MO} = - \frac{\overline{\theta} u_{\star}^3}{\kappa g [\overline{w^{\prime}\theta^{\prime}} + \overline{w^{\prime}q_{t}^{\prime}}]} = - \frac{\overline{\theta} u_{\star}^3}{\kappa g [\frac{J}{\rho c_{pm}} + \theta\frac{D}{\rho L_{v}}]}
```

NK use volume-averaged quantities$\langle X \rangle$ such that 
```math 
\langle X \rangle = \frac{1}{\Delta z} \int_{0}^{\Delta z} X(z) dz
```

> In the `ClimateMachine` DG discretization we interpret these as the nodal variable values at a given model reference altitude $\Delta z$. This value is currently taken as the first interior grid point in the model column description, which typically is of $O(1-10)$ in LES configurations and $O(10-100)$ in the GCM configurations. An alternative to this is obtained by fixing the reference $\Delta z$ and interpolating the DG polynomials to determine variable values at the specified model height. Applying _NK_ equations (17), (18) to the `ClimateMachine` equations we arrive at 

Reviewer: I'm confused by this statement. In NK, the use of volume-averaged quantities leads to modified similarity functions. If we use nodal values (interpreted as point values), the NK approach does not apply, and we should use other standard similarity functions (e.g., those in GLGS). It would be good to clarify what we are doing.
